File: android/src/main/kotlin/com/juandpt/flutter_mediapipe_chat/ModelConfig.kt
package com.juandpt.flutter_mediapipe_chat

data class ModelConfig(
    val path: String,
    val temperature: Float = 0.8f,
    val maxTokens: Int = 1024,
    val topK: Int = 40,
    val randomSeed: Int = 0,
    val loraPath: String? = null,
    val supportedLoraRanks: List<Int>? = null
) {
    companion object {
        fun fromMap(map: Map<*, *>): ModelConfig {
            val path = map["path"] as? String ?: throw IllegalArgumentException("Missing 'path'")
            val temperature = (map["temperature"] as? Number)?.toFloat() ?: 0.8f
            val maxTokens = (map["maxTokens"] as? Number)?.toInt() ?: 1024
            val topK = (map["topK"] as? Number)?.toInt() ?: 40
            val randomSeed = (map["randomSeed"] as? Number)?.toInt() ?: 0
            val loraPath = map["loraPath"] as? String
            val supportedLoraRanks = (map["supportedLoraRanks"] as? List<*>)?.mapNotNull { (it as? Number)?.toInt() }
            return ModelConfig(path, temperature, maxTokens, topK, randomSeed, loraPath, supportedLoraRanks)
        }
    }
}

File: lib/model_config.dart
part of 'flutter_mediapipe_chat.dart';

/// A configuration class for the inference model.
///
/// Contains parameters such as:
/// - [path]: The path to the model file.
/// - [temperature]: Controls the creativity of the response.
/// - [maxTokens]: Maximum number of tokens to generate.
/// - [topK]: Sampling parameter for response generation.
/// - [randomSeed]: Seed for randomness.
/// - [loraPath] and [supportedLoraRanks]: Optional parameters for fine-tuning.
class ModelConfig {
  final String path;
  final double temperature;
  final int maxTokens;
  final int topK;
  final int randomSeed;
  final String? loraPath;
  final List<int>? supportedLoraRanks;

  ModelConfig({
    required this.path,
    this.temperature = 0.8,
    this.maxTokens = 1024,
    this.topK = 40,
    this.randomSeed = 0,
    this.loraPath,
    this.supportedLoraRanks,
  });

  /// Converts this [ModelConfig] into a [Map] for use with native code.
  Map<String, dynamic> toMap() {
    final map = <String, dynamic>{
      'path': path,
      'temperature': temperature,
      'maxTokens': maxTokens,
      'topK': topK,
      'randomSeed': randomSeed,
    };
    if (loraPath != null) {
      map['loraPath'] = loraPath;
    }
    if (supportedLoraRanks != null) {
      map['supportedLoraRanks'] = supportedLoraRanks;
    }
    return map;
  }
}
